File: packages/federation/src/anonymizer.ts
/**
 * On-device anonymization for federated learning.
 *
 * Orchestrates the anonymization pipeline but delegates actual privacy
 * operations (Laplace noise injection, PII stripping, emotion bucketing)
 * to the Rust layer via napiFederatedAnonymize from @aix/core.
 *
 * Privacy invariants (ADR-023, non-negotiable):
 * - No raw user data ever leaves the device.
 * - Patterns anonymized BEFORE leaving device.
 * - Emotion valence bucketed into 5 levels.
 * - Laplace noise epsilon=1.0 on urgency/satisfaction.
 * - No speaker embeddings federated.
 * - Minimum 1000-user aggregation threshold.
 *
 * Maps to rlmx-federation anonymizer.rs.
 */

import type { LifeDomain } from '@aix/shared';
import type { NativeBindings, UnavailableResult } from '@aix/core';
import type {
  AnonymizedPattern,
  Contribution,
  EmotionBucket,
  LoraDelta,
} from './contribution.js';
import { createContribution } from './contribution.js';
import { anonymizationFailed, privacyViolation } from './errors.js';

// ---------------------------------------------------------------------------
// Configuration
// ---------------------------------------------------------------------------

/** Configuration for the anonymization pipeline. */
export interface AnonymizationConfig {
  /** Differential privacy epsilon parameter. Must be 1.0 per ADR-023. */
  epsilon: number;
  /** Minimum number of contributors before patterns can be aggregated. */
  aggregationThreshold: number;
  /** Maximum patterns per contribution to limit information leakage. */
  maxPatternsPerContribution: number;
}

/** Default anonymization config enforcing privacy invariants. */
export const DEFAULT_ANONYMIZATION_CONFIG: Readonly<AnonymizationConfig> = {
  epsilon: 1.0,
  aggregationThreshold: 1000,
  maxPatternsPerContribution: 500,
};

// ---------------------------------------------------------------------------
// Voice-enriched pattern (pre-anonymization input)
// ---------------------------------------------------------------------------

/** A voice-enriched pattern before anonymization. Contains raw data. */
export interface VoiceEnrichedPattern {
  /** Unique identifier. */
  id: string;
  /** Raw query embedding (will be PII-stripped). */
  queryEmbedding: number[];
  /** Actions taken during the interaction. */
  actionsTaken: string[];
  /** Outcome quality [0.0, 1.0]. */
  resultQuality: number;
  /** Raw emotion valence (will be bucketed). */
  voiceTriggerEmotion: number | null;
  /** Raw urgency level (will have Laplace noise applied). */
  urgencyLevel: number;
  /** Interaction modality. */
  interactionModality: 'Voice' | 'Text' | 'Multimodal';
  /** Raw response satisfaction (will have Laplace noise applied). */
  responseSatisfaction: number | null;
  /** Timestamp (ISO 8601). */
  timestamp: string;
}

// ---------------------------------------------------------------------------
// FederationAnonymizer
// ---------------------------------------------------------------------------

/**
 * Federation-level anonymizer that orchestrates the anonymization pipeline.
 *
 * Delegates actual Laplace noise injection to the Rust native layer
 * via napiFederatedAnonymize. When native is unavailable, falls back
 * to a pure-TS implementation that applies the same privacy guarantees.
 */
export class FederationAnonymizer {
  private readonly config: AnonymizationConfig;
  private readonly nativeBindings: NativeBindings | null;

  constructor(
    config: Partial<AnonymizationConfig> = {},
    nativeBindings: NativeBindings | null = null,
  ) {
    this.config = { ...DEFAULT_ANONYMIZATION_CONFIG, ...config };
    this.nativeBindings = nativeBindings;
  }

  /** Return the aggregation threshold. */
  get aggregationThreshold(): number {
    return this.config.aggregationThreshold;
  }

  /** Return the epsilon value. */
  get epsilon(): number {
    return this.config.epsilon;
  }

  /**
   * Anonymize a batch of voice-enriched patterns into a contribution.
   *
   * This is the main entry point for on-device anonymization before federation.
   * Applies the full pipeline: strip PII, bucket emotion, add Laplace noise.
   */
  async anonymizePatterns(
    patterns: VoiceEnrichedPattern[],
    userPseudonym: string,
    domain: LifeDomain,
    loraDelta: LoraDelta | null = null,
  ): Promise<Contribution> {
    if (patterns.length === 0) {
      throw anonymizationFailed('no patterns to anonymize');
    }

    // Limit the number of patterns to prevent information leakage.
    const limit = Math.min(
      this.config.maxPatternsPerContribution,
      patterns.length,
    );
    const selected = patterns.slice(0, limit);

    // Attempt native anonymization first, fall back to TS implementation.
    const anonymized = await this.anonymizeViaBackend(selected);

    return createContribution(userPseudonym, domain, anonymized, loraDelta);
  }

  /**
   * Anonymize patterns using the native backend if available,
   * otherwise fall back to the pure-TS implementation.
   */
  private async anonymizeViaBackend(
    patterns: VoiceEnrichedPattern[],
  ): Promise<AnonymizedPattern[]> {
    if (this.nativeBindings) {
      try {
        const input = JSON.stringify(patterns);
        const result = await this.nativeBindings.napiFederatedAnonymize(input);

        // Check for degraded mode.
        if (
          typeof result === 'object' &&
          result !== null &&
          'status' in result &&
          (result as UnavailableResult).status === 'unavailable'
        ) {
          return this.anonymizeFallback(patterns);
        }

        return JSON.parse(result as string) as AnonymizedPattern[];
      } catch {
        // Native call failed, fall back to TS implementation.
        return this.anonymizeFallback(patterns);
      }
    }

    return this.anonymizeFallback(patterns);
  }

  /**
   * Pure-TS fallback anonymization that applies the same privacy guarantees.
   * Used when the native Rust layer is unavailable.
   */
  private anonymizeFallback(
    patterns: VoiceEnrichedPattern[],
  ): AnonymizedPattern[] {
    return patterns.map((p) => ({
      id: p.id,
      sanitizedEmbedding: FederationAnonymizer.stripPii(p.queryEmbedding),
      actionsTaken: p.actionsTaken,
      resultQuality: p.resultQuality,
      emotionBucket:
        p.voiceTriggerEmotion !== null
          ? FederationAnonymizer.bucketEmotion(p.voiceTriggerEmotion)
          : null,
      noisyUrgency: this.addLaplaceNoise(p.urgencyLevel),
      noisySatisfaction:
        p.responseSatisfaction !== null
          ? this.addLaplaceNoise(p.responseSatisfaction)
          : null,
      interactionModality: p.interactionModality,
      timestamp: p.timestamp,
    }));
  }

  /**
   * Strip PII from an embedding by zeroing the first 4 components
   * (which typically encode speaker-specific information).
   */
  static stripPii(embedding: number[]): number[] {
    const result = [...embedding];
    // Zero out first 4 dimensions (speaker identity components).
    for (let i = 0; i < Math.min(4, result.length); i++) {
      result[i] = 0;
    }
    return result;
  }

  /**
   * Bucket an emotion valence into 5 discrete levels.
   * Privacy-safe: collapses continuous values into coarse categories.
   */
  static bucketEmotion(valence: number): EmotionBucket {
    if (valence < -0.6) return 'VeryNegative';
    if (valence < -0.2) return 'Negative';
    if (valence <= 0.2) return 'Neutral';
    if (valence <= 0.6) return 'Positive';
    return 'VeryPositive';
  }

  /**
   * Add Laplace noise for differential privacy (epsilon from config).
   *
   * Laplace distribution with scale = 1/epsilon. Uses the inverse CDF
   * method: sample U ~ Uniform(-0.5, 0.5), then X = -b * sign(U) * ln(1 - 2|U|).
   */
  addLaplaceNoise(value: number): number {
    const b = 1.0 / this.config.epsilon;
    const u = Math.random() - 0.5;
    const noise = -b * Math.sign(u) * Math.log(1 - 2 * Math.abs(u));
    return value + noise;
  }

  /**
   * Check whether a text string likely contains PII.
   *
   * Simple heuristic: looks for patterns resembling email addresses,
   * phone numbers, or social security numbers.
   */
  static likelyContainsPii(text: string): boolean {
    // Email pattern.
    if (text.includes('@') && text.includes('.')) {
      return true;
    }
    // Phone-like sequences (9+ consecutive digits possibly with separators).
    const digitCount = text.split('').filter((c) => /\d/.test(c)).length;
    if (digitCount >= 9) {
      return true;
    }
    // SSN pattern (NNN-NN-NNNN).
    if (
      text.length === 11 &&
      /^\d{3}-\d{2}-\d{4}$/.test(text)
    ) {
      return true;
    }
    return false;
  }

  /**
   * Validate that an anonymized pattern does not violate privacy invariants.
   * Throws FederationError if PII is detected in action strings.
   */
  static validateAnonymized(pattern: AnonymizedPattern): void {
    for (const action of pattern.actionsTaken) {
      if (FederationAnonymizer.likelyContainsPii(action)) {
        throw privacyViolation(
          'PII detected in anonymized pattern actions',
        );
      }
    }
  }
}

File: crates/rlmx-federation/src/anonymizer.rs
//! On-device anonymization for federated learning.
//!
//! Wraps `rlmx-cognitive`'s `FederatedAnonymizer` and adds federation-specific
//! validation: aggregation threshold enforcement, PII detection, and
//! contribution-level anonymization.
//!
//! Privacy invariants (ADR-023, non-negotiable):
//! - No raw user data ever leaves the device.
//! - Patterns anonymized BEFORE leaving device.
//! - Emotion valence bucketed into 5 levels.
//! - Laplace noise epsilon=1.0 on urgency/satisfaction.
//! - No speaker embeddings federated.
//! - Minimum 1000-user aggregation threshold.

use rlmx_cognitive::sona::LoraDelta;
use rlmx_cognitive::voice_patterns::{
    AnonymizedPattern, FederatedAnonymizer, VoiceEnrichedPattern,
};
use rlmx_kernel::LifeDomain;
use tracing::{debug, warn};

use crate::contribution::Contribution;

/// Configuration for the anonymization pipeline.
#[derive(Debug, Clone)]
pub struct AnonymizationConfig {
    /// Differential privacy epsilon parameter.
    pub epsilon: f64,
    /// Minimum number of contributors before patterns can be aggregated.
    pub aggregation_threshold: usize,
    /// Maximum patterns per contribution to limit information leakage.
    pub max_patterns_per_contribution: usize,
}

impl Default for AnonymizationConfig {
    fn default() -> Self {
        Self {
            epsilon: 1.0,
            aggregation_threshold: 1000,
            max_patterns_per_contribution: 500,
        }
    }
}

/// Federation-level anonymizer that wraps the cognitive crate's `FederatedAnonymizer`
/// and adds contribution packaging.
pub struct FederationAnonymizer {
    inner: FederatedAnonymizer,
    config: AnonymizationConfig,
}

impl std::fmt::Debug for FederationAnonymizer {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("FederationAnonymizer")
            .field("epsilon", &self.config.epsilon)
            .field("config", &self.config)
            .finish()
    }
}

impl FederationAnonymizer {
    /// Create a new federation anonymizer with the given config.
    pub fn new(config: AnonymizationConfig) -> Self {
        let inner = FederatedAnonymizer::new(config.epsilon);
        Self { inner, config }
    }

    /// Return the aggregation threshold.
    pub fn aggregation_threshold(&self) -> usize {
        self.config.aggregation_threshold
    }

    /// Anonymize a batch of voice-enriched patterns into a contribution.
    ///
    /// This is the main entry point for on-device anonymization before federation.
    /// Applies the full pipeline: strip PII, bucket emotion, add Laplace noise.
    pub fn anonymize_patterns(
        &self,
        patterns: &[VoiceEnrichedPattern],
        user_pseudonym: String,
        domain: LifeDomain,
        lora_delta: Option<LoraDelta>,
    ) -> crate::error::Result<Contribution> {
        if patterns.is_empty() {
            return Err(crate::error::FederationError::AnonymizationFailed(
                "no patterns to anonymize".into(),
            ));
        }

        // Limit the number of patterns to prevent information leakage.
        let limit = self
            .config
            .max_patterns_per_contribution
            .min(patterns.len());
        let selected = &patterns[..limit];

        debug!(
            pattern_count = selected.len(),
            domain = ?domain,
            "anonymizing patterns for federation"
        );

        let anonymized: Vec<AnonymizedPattern> = selected
            .iter()
            .map(|p| self.inner.anonymize_pattern(p))
            .collect();

        Ok(Contribution::new(
            user_pseudonym,
            domain,
            anonymized,
            lora_delta,
        ))
    }

    /// Strip PII from a single embedding (delegates to cognitive crate).
    pub fn strip_pii(embedding: &[f32]) -> Vec<f32> {
        FederatedAnonymizer::strip_pii(embedding)
    }

    /// Bucket an emotion valence into 5 discrete levels (delegates to cognitive crate).
    pub fn bucket_emotion(valence: f32) -> rlmx_cognitive::voice_patterns::EmotionBucket {
        FederatedAnonymizer::bucket_emotion(valence)
    }

    /// Add Laplace noise for differential privacy.
    pub fn add_laplace_noise(&self, value: f32) -> f32 {
        self.inner.add_laplace_noise(value)
    }

    /// Check whether a text string likely contains PII.
    ///
    /// Simple heuristic: looks for patterns resembling email addresses,
    /// phone numbers, or social security numbers.
    pub fn likely_contains_pii(text: &str) -> bool {
        // Email pattern
        if text.contains('@') && text.contains('.') {
            return true;
        }
        // Phone-like sequences (10+ consecutive digits possibly with separators)
        let digit_count = text.chars().filter(|c| c.is_ascii_digit()).count();
        if digit_count >= 9 {
            return true;
        }
        // SSN pattern (NNN-NN-NNNN)
        if text.len() == 11
            && text.chars().enumerate().all(|(i, c)| {
                if i == 3 || i == 6 {
                    c == '-'
                } else {
                    c.is_ascii_digit()
                }
            })
        {
            return true;
        }
        false
    }

    /// Validate that an anonymized pattern does not violate privacy invariants.
    pub fn validate_anonymized(pattern: &AnonymizedPattern) -> crate::error::Result<()> {
        // Ensure actions don't contain PII.
        for action in &pattern.actions_taken {
            if Self::likely_contains_pii(action) {
                warn!(action = %action, "PII detected in action string");
                return Err(crate::error::FederationError::PrivacyViolation(
                    "PII detected in anonymized pattern actions".into(),
                ));
            }
        }
        Ok(())
    }
}

impl Default for FederationAnonymizer {
    fn default() -> Self {
        Self::new(AnonymizationConfig::default())
    }
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use chrono::Utc;
    use rlmx_cognitive::voice_patterns::Modality;
    use uuid::Uuid;

    fn make_voice_pattern(emotion: Option<f32>, urgency: f32) -> VoiceEnrichedPattern {
        VoiceEnrichedPattern {
            id: Uuid::new_v4(),
            query_embedding: vec![0.1; 64],
            actions_taken: vec!["test_action".into()],
            result_quality: 0.85,
            voice_trigger_emotion: emotion,
            urgency_level: urgency,
            interaction_modality: Modality::Voice,
            response_satisfaction: Some(0.7),
            timestamp: Utc::now(),
        }
    }

    #[test]
    fn test_anonymize_patterns() {
        let anon = FederationAnonymizer::default();
        let patterns = vec![make_voice_pattern(Some(0.5), 0.3)];
        let contrib = anon
            .anonymize_patterns(&patterns, "pseudo".into(), LifeDomain::Finance, None)
            .unwrap();
        assert_eq!(contrib.pattern_count(), 1);
        assert_eq!(contrib.domain, LifeDomain::Finance);
        assert_eq!(contrib.user_pseudonym, "pseudo");
    }

    #[test]
    fn test_anonymize_empty_patterns_fails() {
        let anon = FederationAnonymizer::default();
        let result = anon.anonymize_patterns(&[], "pseudo".into(), LifeDomain::Health, None);
        assert!(result.is_err());
    }

    #[test]
    fn test_anonymize_respects_max_limit() {
        let config = AnonymizationConfig {
            max_patterns_per_contribution: 2,
            ..Default::default()
        };
        let anon = FederationAnonymizer::new(config);
        let patterns: Vec<_> = (0..5).map(|_| make_voice_pattern(None, 0.5)).collect();
        let contrib = anon
            .anonymize_patterns(&patterns, "pseudo".into(), LifeDomain::Shopping, None)
            .unwrap();
        assert_eq!(contrib.pattern_count(), 2);
    }

    #[test]
    fn test_likely_contains_pii_email() {
        assert!(FederationAnonymizer::likely_contains_pii(
            "user@example.com"
        ));
    }

    #[test]
    fn test_likely_contains_pii_phone() {
        assert!(FederationAnonymizer::likely_contains_pii(
            "call 555-123-4567"
        ));
    }

    #[test]
    fn test_likely_contains_pii_ssn() {
        assert!(FederationAnonymizer::likely_contains_pii("123-45-6789"));
    }

    #[test]
    fn test_no_pii_in_clean_text() {
        assert!(!FederationAnonymizer::likely_contains_pii(
            "pay electric bill"
        ));
    }

    #[test]
    fn test_validate_anonymized_clean() {
        use rlmx_cognitive::voice_patterns::AnonymizedPattern;
        let pattern = AnonymizedPattern {
            id: Uuid::new_v4(),
            sanitized_embedding: vec![0.1; 8],
            actions_taken: vec!["bill_pay".into()],
            result_quality: 0.8,
            emotion_bucket: None,
            noisy_urgency: 0.5,
            noisy_satisfaction: None,
            interaction_modality: Modality::Text,
            timestamp: Utc::now(),
        };
        assert!(FederationAnonymizer::validate_anonymized(&pattern).is_ok());
    }

    #[test]
    fn test_validate_anonymized_with_pii_fails() {
        use rlmx_cognitive::voice_patterns::AnonymizedPattern;
        let pattern = AnonymizedPattern {
            id: Uuid::new_v4(),
            sanitized_embedding: vec![0.1; 8],
            actions_taken: vec!["email john@example.com".into()],
            result_quality: 0.8,
            emotion_bucket: None,
            noisy_urgency: 0.5,
            noisy_satisfaction: None,
            interaction_modality: Modality::Text,
            timestamp: Utc::now(),
        };
        assert!(FederationAnonymizer::validate_anonymized(&pattern).is_err());
    }
}
